api/app/src/main/kotlin/packit/service/RoleService.kt

```diff
@@ -0,0 +1,98 @@
+package packit.service
+
+import org.springframework.http.HttpStatus
+import org.springframework.security.core.GrantedAuthority
+import org.springframework.security.core.authority.SimpleGrantedAuthority
+import org.springframework.stereotype.Service
+import packit.exceptions.PackitException
+import packit.model.Role
+import packit.model.RolePermission
+import packit.repository.RoleRepository
+
+interface RoleService
+{
+    fun getUsernameRole(username: String): Role
+    fun getAdminRole(): Role
+    fun saveRole(roleName: String)
+    fun checkMatchingRoles(rolesToCheck: List<String>): List<Role>
+    fun getGrantedAuthorities(roles: List<Role>): MutableList<GrantedAuthority>
+}
+
+@Service
+class BaseRoleService(
+    private val roleRepository: RoleRepository
+) : RoleService
+{
+    override fun getUsernameRole(username: String): Role
+    {
+        val userRole = roleRepository.findByName(username)
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = username))
+    }
+
+    override fun getAdminRole(): Role
+    {
+        val userRole = roleRepository.findByName("ADMIN")
+
+        if (userRole != null)
+        {
+            return userRole
+        }
+        return roleRepository.save(Role(name = "ADMIN"))
+    }
+
+    override fun saveRole(roleName: String)
+    {
+        if (roleRepository.existsByName(roleName))
+        {
+            throw PackitException("roleAlreadyExists")
+        }
+        val role = Role(name = roleName)
+        roleRepository.save(role)
+    }
+
+    override fun checkMatchingRoles(rolesToCheck: List<String>): List<Role>
+    {
+        val allRoles = roleRepository.findAll()
+        val foundRoles = rolesToCheck.mapNotNull { name -> allRoles.find { it.name == name } }
+
+        if (foundRoles.size != rolesToCheck.size)
+        {
+            throw PackitException("invalidRolesProvided", HttpStatus.BAD_REQUEST)
+        }
+        return foundRoles
+    }
+
+    /**
+     * Authorizes constructed as combination of role names and permission names.
```

Comment:
```suggestion
     * Authorities constructed as combination of role names and permission names.
```

I think authorities can just be permissions can't they? Roles don't have any authorities outside the permissions they provide.

Reply:
yeah that is true but i added them for flexibility if we want to add cases for special roles in the future

Reply:
I'm pretty sure we'd do those as permissions... but I guess it's fine to keep the roles here, might be useful diagnostically at some point.